core/src/test/java/com/graphhopper/routing/weighting/custom/SpeedCalculatorTest.java

```diff
@@ -0,0 +1,184 @@
+        }
+        // negative values are not allowed
+        try {
+            map.put("*", -5);
+            assertEquals(60, calcSpeed(edge, model));
+            fail();
+        } catch (IllegalArgumentException e) {
+            assertTrue(e.getMessage().contains("max_speed.road_class cannot be smaller than 0.0, was -5.0"), e.getMessage());
+        }
+    }
+
+    @Test
+    public void maxSpeedMultiple() {
+        EnumEncodedValue<RoadClass> roadClass = em.getEnumEncodedValue(RoadClass.KEY, RoadClass.class);
+        EnumEncodedValue<RoadEnvironment> roadEnvironment = em.getEnumEncodedValue(RoadEnvironment.KEY, RoadEnvironment.class);
+        edge.set(roadClass, RoadClass.PRIMARY);
+        edge.set(roadEnvironment, RoadEnvironment.BRIDGE);
+        Map<String, Object> roadClassMap = new HashMap<>();
+        roadClassMap.put(RoadClass.PRIMARY.toString(), 40);
+        Map<String, Object> roadEnvironmentMap = new HashMap<>();
+        roadEnvironmentMap.put(RoadEnvironment.BRIDGE.toString(), 20);
+        CustomModel model = new CustomModel();
+        model.getMaxSpeed().put(RoadClass.KEY, roadClassMap);
+        model.getMaxSpeed().put(RoadEnvironment.KEY, roadEnvironmentMap);
+        // the more restrictive max speed 'wins'
+        assertEquals(20, calcSpeed(edge, model));
+    }
+
+    @Test
+    public void speedFactor() {
+        // here we use speed_factor to adjust speed
+        CustomModel model = new CustomModel();
+        Map<String, Object> map = new HashMap<>();
+        map.put("*", 0.1);
+        model.getSpeedFactor().put(RoadClass.KEY, map);
+        assertEquals(6, calcSpeed(edge, model));
+        // our edge might or might not be affected
+        EnumEncodedValue<RoadClass> roadClass = em.getEnumEncodedValue(RoadClass.KEY, RoadClass.class);
+        edge.set(roadClass, RoadClass.PRIMARY);
+        model.getSpeedFactor().clear();
+        assertEquals(60, calcSpeed(edge, new CustomModel()));
+        map.clear();
+        map.put(RoadClass.PRIMARY.toString(), 0.2);
+        model.getSpeedFactor().put(RoadClass.KEY, map);
+        assertEquals(12, calcSpeed(edge, model));
+    }
+
+    @Test
+    public void invalidSpeedFactor() {
+        CustomModel model = new CustomModel();
+        Map<String, Object> map = new HashMap<>();
+        map.put("*", 1.1);
+        model.getSpeedFactor().put(RoadClass.KEY, map);
+        try {
+            assertEquals(6, calcSpeed(edge, model));
+            fail();
+        } catch (IllegalArgumentException e) {
+            assertTrue(e.getMessage().contains("speed_factor.road_class cannot be bigger than 1.0, was 1.1"), e.getMessage());
+        }
+    }
+
+    @Test
+    public void speedFactorMultiple() {
+        // use speed_factor to adjust speed
+        // todonow: check invalid values
+        // todonow: what if we use "car.average_speed" as key?
```

Comment:
@karussell

Reply:
> todonow: what if we use "car.average_speed" as key? 

You mean something like `speed_factor: car.average_speed: "<30": 0.5`? This should not cause any problems although it is not something that we explicitly test for.